File: src/Raven.Server/Documents/Patch/PatchDebugActions.cs
using Sparrow.Json.Parsing;

namespace Raven.Server.Documents.Patch
{
    public class PatchDebugActions
    {
        public readonly DynamicJsonArray LoadDocument = new DynamicJsonArray();
        public readonly DynamicJsonArray PutDocument = new DynamicJsonArray();
        public readonly DynamicJsonArray DeleteDocument = new DynamicJsonArray();
       
        public readonly DynamicJsonArray GetCounter = new DynamicJsonArray();
        public readonly DynamicJsonArray IncrementCounter = new DynamicJsonArray();
        public readonly DynamicJsonArray DeleteCounter = new DynamicJsonArray();
        
        public readonly DynamicJsonArray GetTimeSeries = new DynamicJsonArray();
        public readonly DynamicJsonArray AppendTimeSeries = new DynamicJsonArray();
        public readonly DynamicJsonArray IncrementTimeSeries = new DynamicJsonArray();
        public readonly DynamicJsonArray DeleteTimeSeries = new DynamicJsonArray();
     
        public DynamicJsonValue GetDebugActions()
        {
            return new DynamicJsonValue
            {
                [nameof(LoadDocument)] = new DynamicJsonArray(LoadDocument.Items),
                [nameof(PutDocument)] = new DynamicJsonArray(PutDocument.Items),
                [nameof(DeleteDocument)] = new DynamicJsonArray(DeleteDocument.Items),

                [nameof(GetCounter)] = new DynamicJsonArray(GetCounter.Items),
                [nameof(IncrementCounter)] = new DynamicJsonArray(IncrementCounter.Items),
                [nameof(DeleteCounter)] = new DynamicJsonArray(DeleteCounter.Items),
                  
                [nameof(GetTimeSeries)] = new DynamicJsonArray(GetTimeSeries.Items),
                [nameof(AppendTimeSeries)] = new DynamicJsonArray(AppendTimeSeries.Items),
                [nameof(IncrementTimeSeries)] = new DynamicJsonArray(IncrementTimeSeries.Items),
                [nameof(DeleteTimeSeries)] = new DynamicJsonArray(DeleteTimeSeries.Items)
            };
        }

        public void Clear()
        {
            LoadDocument.Clear();
            PutDocument.Clear();
            DeleteDocument.Clear();
            
            GetCounter.Clear();
            IncrementCounter.Clear();
            DeleteCounter.Clear();
            
            GetTimeSeries.Clear();
            AppendTimeSeries.Clear();
            IncrementTimeSeries.Clear();
            DeleteTimeSeries.Clear();
        }
    }
}

File: src/Raven.Studio/typescript/viewmodels/database/patch/patchDebugActions.ts
import { highlight, languages } from "prismjs";


class patchDebugActions {
    loadDocument = ko.observableArray<any>();
    putDocument = ko.observableArray<any>();
    deleteDocument = ko.observableArray<string>();
    getCounter = ko.observableArray<any>();
    incrementCounter = ko.observableArray<any>();
    deleteCounter = ko.observableArray<string>();
    getTimeSeries = ko.observableArray<any>();
    appendTimeSeries = ko.observableArray<any>();
    deleteTimeSeries = ko.observableArray<any>();

    loadedCount: KnockoutComputed<string>;
    modifiedCount: KnockoutComputed<string>;
    deletedCount: KnockoutComputed<string>;

    showDocumentsInModified = ko.observable<boolean>(false);
    showTimeSeriesValuesInLoaded = ko.observable<boolean>(false);
    showTimeSeriesValuesInModified = ko.observable<boolean>(false);
    
    constructor() {
        this.initObservables();
    }

    formatAsJson(input: KnockoutObservable<any> | any) {
        return ko.pureComputed(() => {
            const value = ko.unwrap(input);
            if (value === undefined) {
                return "";
            } else {
                const json = JSON.stringify(value, null, 4);
                return highlight(json, languages.javascript, "js");
            }
        });
    }
    
    private initObservables() {
        this.loadedCount = ko.pureComputed(() => {
            const totalLoadedCount = this.loadDocument().length +
                this.getCounter().length +
                this.getTimeSeries().length;

            return totalLoadedCount ? totalLoadedCount.toLocaleString() : "";
        });

        this.modifiedCount = ko.pureComputed(() => {
            const totalModifiedCount = this.putDocument().length +
                this.incrementCounter().length +
                this.appendTimeSeries().length;

            return totalModifiedCount ? totalModifiedCount.toLocaleString() : "";
        });

        this.deletedCount = ko.pureComputed(() => {
            const totalDeletedCount = this.deleteDocument().length +
                this.deleteCounter().length +
                this.deleteTimeSeries().length;

            return totalDeletedCount ? totalDeletedCount.toLocaleString() : "";
        });
    }
    
    fill(incomingDto: Raven.Server.Documents.Patch.PatchDebugActions) {
        this.loadDocument(incomingDto.LoadDocument);
        this.putDocument(incomingDto.PutDocument);
        this.deleteDocument(incomingDto.DeleteDocument);

        this.getCounter(incomingDto.GetCounter);
        this.incrementCounter(incomingDto.IncrementCounter);
        this.deleteCounter(incomingDto.DeleteCounter);

        this.getTimeSeries(incomingDto.GetTimeSeries);
        this.appendTimeSeries(incomingDto.AppendTimeSeries);
        this.deleteTimeSeries(incomingDto.DeleteTimeSeries);
    }
    
    reset() {
        this.loadDocument([]);
        this.putDocument([]);
        this.deleteDocument([]);

        this.getCounter([]);
        this.incrementCounter([]);
        this.deleteCounter([]);

        this.getTimeSeries([]);
        this.appendTimeSeries([]);
        this.deleteTimeSeries([]);
    }
    
}

export = patchDebugActions;
